quinn-proto/src/cid_queue.rs

```diff
@@ -0,0 +1,199 @@
+use std::ops::Range;
+
+use crate::{
+    shared::{IssuedCid, ResetToken},
+    ConnectionId,
+};
+
+/// Sliding window of active Connection IDs
+///
+/// May contain gaps due to packet loss or reordering
+#[derive(Debug)]
+pub struct CidQueue {
+    /// Ring buffer indexed by `self.cursor`
+    buffer: [Option<(ConnectionId, ResetToken)>; Self::LEN],
+    /// Index at which circular buffer addressing is based
+    cursor: usize,
+    /// Sequence number of `self.buffer[cursor]`
+    offset: u64,
+}
+
+impl CidQueue {
+    pub const LEN: usize = 8;
+
+    pub fn new(offset: u64) -> Self {
+        Self {
+            buffer: [None; Self::LEN],
+            cursor: 0,
+            offset,
+        }
+    }
+
+    pub fn insert(&mut self, cid: IssuedCid) -> Result<(), InsertError> {
+        let index = match cid.sequence.checked_sub(self.offset) {
+            None => return Err(InsertError::Retired),
+            Some(x) => x,
+        };
+        if index >= Self::LEN as u64 {
+            return Err(InsertError::ExceedsLimit);
+        }
+        let index = (self.cursor + index as usize) % Self::LEN;
+        self.buffer[index] = Some((cid.id, cid.reset_token));
+        Ok(())
+    }
+
+    /// Returns the possibly-empty range of newly retired CIDs
+    pub fn retire_prior_to(&mut self, sequence: u64) -> Range<u64> {
+        let n = match sequence.checked_sub(self.offset) {
+            None => return 0..0,
+            Some(n) => n as usize,
+        };
+        for i in 0..n {
+            self.buffer[(self.cursor + i) % Self::LEN] = None;
+        }
+        let orig_offset = self.offset;
+        self.offset = sequence;
+        self.cursor = (self.cursor + n) % Self::LEN;
+        orig_offset..sequence
+    }
+
+    /// Returns a new CID if any were available and a possibly-empty range preceding it to retire
+    pub fn next(&mut self) -> Option<(IssuedCid, Range<u64>)> {
+        for i in 0..Self::LEN {
+            let index = (self.cursor + i) % Self::LEN;
+            if let Some((id, reset_token)) = self.buffer[index].take() {
+                let orig_offset = self.offset;
+                self.offset += i as u64 + 1;
+                self.cursor = (self.cursor + i + 1) % Self::LEN;
+                let sequence = orig_offset + i as u64;
+                return Some((
```

Comment:
Another style nit: I think the formatted version will be nicer if you bind the intermediate `IssuedCid` here.